interface/task_executor.py

```diff
@@ -0,0 +1,58 @@
+"""
+This module defines public interface from TaskExecutor class family.
+Also defines interface that basecaller dependency injection MUST implement.
+See test/sample_app.py for example usage.
+"""
+
+from dataclasses import dataclass
+from typing import List, Tuple
+
+import numpy as np
+
+from .data_types import FilePath, OutputData
+
+@dataclass
+class BasecallerInterface:
+    """
+    Task executors in this library can use any injected
+    basecaller object that implements following interface.
+    No default basecaller is available.
+    """
+
+    def call_raw_signal(self, signal: np.ndarray) -> Tuple[str, str]:
+        """
+        Method implements signal basecalling. Takes one argument
+        of numpy.ndarray of numpy.float32 values.
+        Please note that input signal is already normalized by library.
+        This can be discussed in the future (see source/task_executor.py).
+
+        Method returns a tuple of strings. Those are basecalled sequence
+        and quality scores (as defined in fastq file format) respectively.
+        """
+        raise NotImplementedError
+
+
+class TaskExecutorInterface:
+
+    def __init__(self) -> None:
+        self.caller = None
+
+    def set_caller(self, caller: BasecallerInterface) -> None:
```

Comment:
shouldn't it rather take caller as constructor param?

Reply:
Well, I'm not sure this is the best design, but I couldn't do better. caller is an external dependency. This library should be able to utilize any basecaller as long as it implements the `BasecallerInterface`. While ArgumentParser provides integrator with created and initialized `TaskExecutor` object, it has no way to know what caller to supply. And I think it shouldn't have. It should be an integrator responsibilty in my opinion to supply a correct caller.

Reply:
Well, that tells you that your whole design is off and has cross-cutting concerns. For example, you can have `execute_task_batch` to also take `caller` as an argument. That way, `TaskExecutor` deals *only* with executing in some fashion and does not care what the caller is. You can then "bind" caller to task executor in another level of abstraction

Reply:
That is true. But once I have `caller` as parameter of `execute_task_batch()` I have to serialize the caller object in parallel execution to communicate the caller object to the workers. If it was a data member, function `_call_read()` could be made a class method in some TaskExecutorBase class other executors would inherit from and caller object could be accessed naturally. (My original plan with this). That way it would be beneficial for performance since serializing large caller objects can be costly.

Reply:
Hmm, ok, now I see this. But then I don't understand how this works if you set caller *after* constructing the pool -- once pool is created, I would assume workers are already spawned and it would be too late to assign caller in the parent process

Reply:
THAT is an excellent question. Haven't thought about it that way.